ir/src/main/scala/com/armanbilge/vilcacora/ir/ModelIR.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2023 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.armanbilge.vilcacora.ir
+
+/** Represents the data type of a tensor's elements and its size in bytes.
+  */
+sealed abstract class DataType { def sizeInBytes: Int }
+object DataType {
+  case object Float32 extends DataType { override def sizeInBytes: Int = 4 }
+  case object Float64 extends DataType { override def sizeInBytes: Int = 8 } // For ONNX DOUBLE
+  case object Float16 extends DataType { override def sizeInBytes: Int = 2 }
+  case object BFloat16 extends DataType { override def sizeInBytes: Int = 2 }
+
+  case object Int64 extends DataType { override def sizeInBytes: Int = 8 }
+  case object Int32 extends DataType { override def sizeInBytes: Int = 4 }
+  case object Int16 extends DataType { override def sizeInBytes: Int = 2 }
+  case object Int8 extends DataType { override def sizeInBytes: Int = 1 }
+
+  case object UInt64 extends DataType { override def sizeInBytes: Int = 8 }
+  case object UInt32 extends DataType { override def sizeInBytes: Int = 4 }
+  case object UInt16 extends DataType { override def sizeInBytes: Int = 2 }
+  case object UInt8 extends DataType { override def sizeInBytes: Int = 1 }
+
+  case object Bool extends DataType { override def sizeInBytes: Int = 1 }
+  // Add other supported types  from ONNX
+}
+
+/** Describes a single, named memory buffer (a tensor).
+  */
+final case class Allocation(
+    name: String,
+    dataType: DataType,
+    shape: List[Int],
+    initialData: Option[Array[Byte]] = None,
+)
+
+/** representation of the `kernel_type` attribute for SVM nodes.
+  */
+sealed trait SVMKernel
+object SVMKernel {
+  case object LINEAR extends SVMKernel
+  case object POLY extends SVMKernel
+  case object RBF extends SVMKernel
+  case object SIGMOID extends SVMKernel
```

Comment:
Some typical style convention for Scala:

```suggestion
sealed abstract class SvmKernel
object SvmKernel {
  case object Linear extends SVMKernel
  case object Poly extends SVMKernel
  case object Rbf extends SVMKernel
  case object Sigmoid extends SVMKernel
```